src/SdkCommon/ClientRuntime/ClientRuntime/RetryAfterDelegatingHandler.cs

```diff
@@ -44,35 +53,72 @@ public RetryAfterDelegatingHandler(DelegatingHandler innerHandler)
+                                //       happening indicating that request has been aborted.
+                                await response.Content?.ReadAsStringAsync();
+
+                                var oldResponse = previousResponseMessage;
+                                previousResponseMessage = response;
+                                oldResponse?.Dispose();
+                            }
+                            catch
+                            {
+                                // We can end up getting errors reading the content of the message if the connection was closed.
+                                // These errors will be ignored and the previous response will continue to be used.
+                                if (previousResponseMessage != null)
+                                {
+                                    response = previousResponseMessage;
+                                }
+                             }
+
+                            try
+                            {
+                                // and we get a number of seconds from the header
+                                string retryValue = response.Headers.GetValues("Retry-After").FirstOrDefault();
+                                var retryAfter = int.Parse(retryValue, CultureInfo.InvariantCulture);
 
-                            // and try again
-                            continue;
-                        } catch {
-                            // if something throws while trying to get the retry-after
-                            // we're just going to suppress it. let the response go
-                            // back to the consumer.
+                                // wait for that duration
+                                await Task.Delay(TimeSpan.FromSeconds(retryAfter), cancellationToken);
+
+                                // and try again
+                                continue;
+                            }
+                            catch
+                            {
+                                // if something throws while trying to get the retry-after
+                                // we're just going to suppress it. let the response go
+                                // back to the consumer.
+                            }
                         }
                     }
                 }
+                catch (TaskCanceledException) when(previousResponseMessage != null)
+                {
+                    // We can get Task Canceled Exceptions while calling the base.SendAsync(...) and
+                    // we do not want to let these bubble out when we have a previous response message to return.
+                }
+
                 // if we haven't hit continue, then return the response up the stream
-                return response;
-            } while( true );
+                return response ?? previousResponseMessage;
```

Comment:
This can still return null, right?

Reply:
We will always have a response or previousResponseMessage here.  It is possible that cancellation occurs before we get any response and in that case the TaskCanceledException will be allowed to bubble up and we will never get here.

Reply:
What happens if the error happened in the `SendAsync` method and that throws something other than `TaskCanceledException`?

Reply:
That exception will bubble up and this line will never be executed.